_source/_posts/2020-11-18-build-a-graphql-nodejs-api.md

```diff
@@ -0,0 +1,303 @@
+            "gender": "Female",
+            "ip_address": "26.58.193.2"
+        },
+        {
+            "id": 2,
+            "first_name": "Giavani",
+            "last_name": "Frediani",
+            "email": "gfrediani1@senate.gov",
+            "gender": "Male",
+            "ip_address": "229.179.4.212"
+        },
+        {
+            "id": 3,
+            "first_name": "Noell",
+            "last_name": "Bea",
+            "email": "nbea2@imageshack.us",
+            "gender": "Female",
+            "ip_address": "180.66.162.255"
+        },
+        {
+            "id": 4,
+            "first_name": "Willard",
+            "last_name": "Valek",
+            "email": "wvalek3@vk.com",
+            "gender": "Male",
+            "ip_address": "67.76.188.26"
+        }
+    ]
+}
+
+module.exports = Object.assign({}, { start });
+```
+
+This file kicks off your Express server with the `start` function. To use GraphQL you must bring in the GraphQL module and tell your application to use it. You can specify whatever endpoint you like. Here you have set it to `/GraphQL`. At the bottom of the file is some sample data. The trick now is to connect the sample data to your `GraphQL` endpoint.
+
+You do this by building a schema. The schema has definitions for your objects and object types. Here you have an `Employees` class, which is an array of `Employee` objects. `Employee` itself has several primitive fields attached to it. You also must define your query. Here is where you tell the client what functions are available to it via the server. In this case, you have an `employees` function that will return the full list of employees and an `employee` function that will return one employee based on a required `id` parameter. In the `root` object, you define handlers for these functions. For `employees`, you are just returning the sample data provided. For `employee`, you must collect the parameters and find the employee's data based on the id property of the query.  
+
+You are also using the `ensureAuthenticated` middleware in the `GraphQL` endpoint. This will call your `jwtVerifier` and check the token presented in the `Authorization` header. Clients that don't provide a valid JWT will be rejected with a `401` (unauthorized) HTTP code.  
+
+Lastly, you need to provide an entry point for your application. Create a new file called `index.js` and add the following code to it:
+
+```javascript
+'use strict'
+
+require('dotenv').config();
+
+const server = require('./server')
+
+const port = process.env.APP_BASE_PORT;
+
+server.start({
+    port: port
+}).then(app => {
+    console.log('Application is now running on port ' + port);
+})
+```
+
+This file does two notable things. First, it calls dotenv's `config()` function. Per the documentation, this should be placed as early in the application as possible. Second, it kicks off the server using the server file you just created.
+
+## Testing your Web API
```

Comment:
I used to get corrected by Lindsay on this one. She'd change it to "Test Your Web API".